internal/gps/source_cache_bolt_encode.go

```diff
@@ -0,0 +1,439 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/golang/dep/internal/gps/pkgtree"
+	"github.com/pkg/errors"
+)
+
+// cacheEncodeUnpairedVersion returns an encoded UnpairedVersion.
+func cacheEncodeUnpairedVersion(uv UnpairedVersion) ([]byte, error) {
+	var pre string
+	switch uv.Type() {
+	case IsBranch:
+		if uv.(branchVersion).isDefault {
+			pre = "defaultBranch:"
+		} else {
+			pre = "branch:"
+		}
+	case IsSemver, IsVersion:
+		pre = "ver:"
+	default:
+		return nil, fmt.Errorf("unrecognized version type: %d", uv.Type())
+	}
+	return []byte(pre + uv.String()), nil
+}
+
+// cacheDecodeUnpairedVersion decodes and returns a new UnpairedVersion.
+func cacheDecodeUnpairedVersion(b []byte) (UnpairedVersion, error) {
+	const br, dbr, ver = "branch:", "defaultBranch:", "ver:"
+	s := string(b)
+	switch {
+	case strings.HasPrefix(s, br):
+		return NewBranch(strings.TrimPrefix(s, br)), nil
+	case strings.HasPrefix(s, dbr):
+		return newDefaultBranch(strings.TrimPrefix(s, dbr)), nil
+	case strings.HasPrefix(s, ver):
+		return NewVersion(strings.TrimPrefix(s, ver)), nil
+	default:
+		return nil, fmt.Errorf("unrecognized prefix: %s", s)
+	}
+}
+
+// cacheDecodeProjectProperties returns a new ProjectRoot and ProjectProperties with the
+// data encoded in the key/value pair.
+func cacheDecodeProjectProperties(k, v []byte) (ProjectRoot, ProjectProperties, error) {
+	var pp ProjectProperties
+	ks := strings.SplitN(string(k), ",", 2)
+	ip := ProjectRoot(ks[0])
+	if len(ks) > 1 {
+		pp.Source = ks[1]
+	}
+	if len(v) == 0 {
+		pp.Constraint = Any()
+	} else {
+		const br, dbr, ver, rev = "branch:", "defaultBranch:", "ver:", "rev:"
+		vs := string(v)
+		switch {
+		case strings.HasPrefix(vs, br):
+			pp.Constraint = NewBranch(strings.TrimPrefix(vs, br))
+
+		case strings.HasPrefix(vs, dbr):
+			pp.Constraint = newDefaultBranch(strings.TrimPrefix(vs, dbr))
+
+		case strings.HasPrefix(vs, ver):
+			vs = strings.TrimPrefix(vs, ver)
+			if c, err := NewSemverConstraint(vs); err != nil {
+				pp.Constraint = NewVersion(vs)
+			} else {
+				pp.Constraint = c
+			}
+
+		case strings.HasPrefix(vs, rev):
+			pp.Constraint = Revision(strings.TrimPrefix(vs, rev))
+
+		default:
+			return "", ProjectProperties{}, fmt.Errorf("unrecognized prefix: %s", vs)
+		}
+	}
+
+	return ip, pp, nil
+}
+
+// cacheEncodeProjectProperties returns a key/value pair containing the encoded
+// ProjectRoot and ProjectProperties.
+func cacheEncodeProjectProperties(ip ProjectRoot, pp ProjectProperties) ([]byte, []byte, error) {
+	k := string(ip)
+	if len(pp.Source) > 0 {
+		k += "," + pp.Source
+	}
+	if pp.Constraint == nil || IsAny(pp.Constraint) {
+		return []byte(k), []byte{}, nil
+	}
+
+	if v, ok := pp.Constraint.(Version); ok {
```

Comment:
Yeah, I modeled it after dep's serialization methods: https://github.com/golang/dep/blob/master/manifest.go#L300

I'd like to be able to just call a method on `Constraint` here. `typedString() string` is functionally close to what we need, but maybe we want a `typedEncoding() []byte` or `typedEncoding() someProtobufOneofType`

Reply:
> Yeah, I modeled it after dep's serialization methods: 

oh for sure, i get why - i'm just personally loathe to do that in gps because i'm accustomed to switching on the underlying types, and the `.Type()` system i wrote for it has always felt less expressive. i don't think it actually is, but it's a gut feel.

> but maybe we want a `typedEncoding() []byte` or `typedEncoding() someProtobufOneofType`

i've not actually used protobufs before (though i have used others); my guess, though, is that we'd end up relying on generated code to create whatever methods we need, rather than writing them ourselves.